### ISSUE TITLE
关于总loss公式

### ISSUE THREAD

**Issue by Jacobxz** [bupt-ai-cz/HHCL-ReID#6]
loss = self.hard_weight * (self.cross_entropy(output_hard, targets) + (1 - self.hard_weight) * self.cross_entropy(output_mean, targets))

您好，我想求教一下，这个loss公式和论文中的公式2吻合吗？

**Comment by Jacobxz** [comment]
看岔了不好意思

**Comment by Jacobxz** [comment]
您好，我问一下当采用cmhybrid时，这种损失函数为什么设计成如下loss = self.hard_weight * (self.cross_entropy(output_hard, targets) + (1 - self.hard_weight) * self.cross_entropy(output_mean, targets))

还有就是，self.hard_weight怎么确定的，其他超参数会影响吗，您是先确定哪一个参数呢？

**Comment by hu-zheng** [comment]
> 您好，我问一下当采用cmhybrid时，这种损失函数为什么设计成如下loss = self.hard_weight * (self.cross_entropy(output_hard, targets) + (1 - self.hard_weight) * self.cross_entropy(output_mean, targets))
> 
> 还有就是，self.hard_weight怎么确定的，其他超参数会影响吗，您是先确定哪一个参数呢？

1. 这类包含多项的损失函数，通常都每项前乘上一个系数，这些系数作为超参数来设置对应项在整体loss中的占比。
2. 本文中的self.hard_weight，只是在其取值区间内通过设置不同的值进行多组实验，取了实验结果中最好的一个设置；通过观察这些实验结果一般只能得到一个大概的趋势，而这个值也并不一定是最优的；另外多个超参数之间肯定是会有影响的，如果要想准确得到超参数的最优解可能需要用到参数搜索的策略。

**Comment by Jacobxz** [comment]
loss = 0.25* self.cross_entropy(output_hard, targets) +0.75 * self.cross_entropy(output_mean, targets)的值低于0* self.cross_entropy(output_hard, targets) +1 * self.cross_entropy(output_mean, targets)也就是cm，按道理应该要高一点，但是实际map结果低于cm，这一般是什么出了问题呢？
        
    
    
     
    
        
                    ***@***.***
        
    
 

    在2022年3月15日 ***@***.***> 写道： 




您好，我问一下当采用cmhybrid时，这种损失函数为什么设计成如下loss = self.hard_weight * (self.cross_entropy(output_hard, targets) + (1 - self.hard_weight) * self.cross_entropy(output_mean, targets))
还有就是，self.hard_weight怎么确定的，其他超参数会影响吗，您是先确定哪一个参数呢？


这类包含多项的损失函数，通常都每项前乘上一个系数，这些系数作为超参数来设置对应项在整体loss中的占比。
本文中的self.hard_weight，只是在其取值区间内通过设置不同的值进行多组实验，取了实验结果中最好的一个设置；通过观察这些实验结果一般只能得到一个大概的趋势，而这个值也并不一定是最优的；另外多个超参数之间肯定是会有影响的，如果要想准确得到超参数的最优解可能需要用到参数搜索的策略。

**Comment by Jacobxz** [comment]
font{
        line-height: 1.6;
    }
    ul,ol{
        padding-left: 20px;
        list-style-position: inside;
    }


    
    
        
            
        
    
    
        一般是先确定哪个超参数？这么晚打扰了
        
    
     
    
        
                    ***@***.***
        
    
 

    在2022年3月15日 ***@***.***> 写道： 




您好，我问一下当采用cmhybrid时，这种损失函数为什么设计成如下loss = self.hard_weight * (self.cross_entropy(output_hard, targets) + (1 - self.hard_weight) * self.cross_entropy(output_mean, targets))
还有就是，self.hard_weight怎么确定的，其他超参数会影响吗，您是先确定哪一个参数呢？


这类包含多项的损失函数，通常都每项前乘上一个系数，这些系数作为超参数来设置对应项在整体loss中的占比。
本文中的self.hard_weight，只是在其取值区间内通过设置不同的值进行多组实验，取了实验结果中最好的一个设置；通过观察这些实验结果一般只能得到一个大概的趋势，而这个值也并不一定是最优的；另外多个超参数之间肯定是会有影响的，如果要想准确得到超参数的最优解可能需要用到参数搜索的策略。

**Comment by hu-zheng** [comment]
手动调参是一件很费时费力、需要经验的事。对于一个参数，有时候可能并不是聊胜于无，可能也并非是越大越好。要确定一个较优的参数确实要做大量实验。
我调参一般是优先确定自己最关心的、影响比较大的参数（比如：我们方法里的self.hard_weight），但这个前提是需要在相对一组对靠谱的超参基础下实验；而对那些不太关心的超参数，可以先参考之前的一些工作，甚至follow他们的配置，如果有必要之后再微调。

**Comment by Jacobxz** [comment]
```
class CM_Ave(autograd.Function):

    @staticmethod
    def forward(ctx, inputs, targets, features, momentum):
        ctx.features = features
        ctx.momentum = momentum
        ctx.save_for_backward(inputs, targets)
        outputs = inputs.mm(ctx.features.t())

        return outputs

    @staticmethod
    def backward(ctx, grad_outputs):
        inputs, targets = ctx.saved_tensors

        grad_inputs = None
        if ctx.needs_input_grad[0]:
            grad_inputs = grad_outputs.mm(ctx.features)

        batch_centers = collections.defaultdict(list)
        for instance_feature, index in zip(inputs, targets.tolist()):
            batch_centers[index].append(instance_feature)

        for index, features in batch_centers.items():
            distances = []
            for feature in features:
                distance = feature.unsqueeze(0).mm(ctx.features[index].unsqueeze(0).t())[0][0]
                distances.append(distance.cpu().numpy())
           
            # 平均
            mean = torch.stack(features, dim=0).mean(0)
            ctx.features[index] = ctx.features[index] * ctx.momentum + (1 - ctx.momentum) * mean
            ctx.features[index] /= ctx.features[index].norm()

        return grad_inputs, None, None, None


def cm_ave(inputs, indexes, features, momentum=0.5):
    return CM_Ave.apply(inputs, indexes, features, torch.Tensor([momentum]).to(inputs.device))
```
![image](https://user-images.githubusercontent.com/48195963/160243053-1c69e314-a00b-40d2-b1c9-c527cc40ea1a.png)
你好，我根据代码复现了一下平均这个情况，代码如上，能不能帮我看看对不对，
还有我用上述代码实验结果偏差有点大，在想是不是代码有问题，能帮我看看吗，麻烦，我的batchsize为128
![image](https://user-images.githubusercontent.com/48195963/160243381-87c70ea9-9ff9-4c4d-bfe6-c7cbef5f5cd4.png)

**Comment by hu-zheng** [comment]
batchsize=128效果可能会略低于256，但是是不会有这么大差别。按我原始的代码把hard_weight设为0也这么低吗？
我看你的这部分代码应该是没有问题，你检查一下features或者momentum初始化对不对，别的地方有没有问题
